acme/http_test.go

```diff
@@ -212,6 +212,131 @@ func TestRetryBackoffArgs(t *testing.T) {
+			responseHttpStatusCode: http.StatusOK,
+			expectedRetryCount:     1,
+		},
+		{
+			desc:                   "does not retry post request that returns response code that is not retriable in default isRetriable configuration",
+			requestMethodType:      http.MethodPost,
+			responseHttpStatusCode: http.StatusUnprocessableEntity,
+			expectedRetryCount:     0,
+		},
+		{
+			desc:                   "retries get request that returns response code that is retriable in default isRetriable configuration",
+			requestMethodType:      http.MethodGet,
+			responseHttpStatusCode: http.StatusTooManyRequests,
+			expectedRetryCount:     1,
+		},
+		{
+			desc:                   "does not retry get request that returns response code that is not retriable in default isRetriable configuration",
+			requestMethodType:      http.MethodGet,
+			responseHttpStatusCode: http.StatusNotFound,
+			expectedRetryCount:     0,
+		},
+		{
+			desc:                   "retries post request that returns response code that is retriable in the client.ShouldRetry configuration",
+			requestMethodType:      http.MethodPost,
+			responseHttpStatusCode: http.StatusInternalServerError,
+			shouldRetryFn: func(httpStatusCode int) bool {
+				return httpStatusCode >= 500
+			},
+			expectedRetryCount: 1,
+		},
+		{
+			desc:                   "does not retry post request that returns response code that is not retriable in the client.ShouldRetry configuration",
+			requestMethodType:      http.MethodPost,
+			responseHttpStatusCode: http.StatusTooManyRequests,
+			shouldRetryFn: func(httpStatusCode int) bool {
+				return false
+			},
+			expectedRetryCount: 0,
+		},
+		{
+			desc:                   "retries get request that returns response code that is retriable  in the client.ShouldRetry configuration",
+			requestMethodType:      http.MethodGet,
+			responseHttpStatusCode: http.StatusTooManyRequests,
+			shouldRetryFn: func(httpStatusCode int) bool {
+				return httpStatusCode == http.StatusTooManyRequests
+			},
+			expectedRetryCount: 1,
+		},
+		{
+			desc:                   "does not retry get request that returns response code that is not retriable in the client.ShouldRetry configuration",
+			requestMethodType:      http.MethodGet,
+			responseHttpStatusCode: http.StatusTooManyRequests,
+			shouldRetryFn: func(httpStatusCode int) bool {
+				return httpStatusCode <= 399 || httpStatusCode >= 500
+			},
+			expectedRetryCount: 0,
+		},
+	}
+
+	for _, test := range tt {
+		t.Run(test.desc, func(t *testing.T) {
+			ts := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+				w.WriteHeader(test.responseHttpStatusCode)
+			}))
+			defer ts.Close()
+
+			ctx, cancel := context.WithCancel(context.Background())
+			var retryNr int
+
+			client := &Client{
+				Key: testKey,
+				RetryBackoff: func(n int, r *http.Request, res *http.Response) time.Duration {
+					retryNr++
+					if retryNr == 1 {
+						cancel()
+					}
```

Comment:
Why only `cancel()` if `retryNr == 1` ?

Reply:
Ah you never go past 1 retry is that it?

Reply:
yeah I just wanted to be able to detect that retried happened, it doesn't matter how many time at least once is enough.

Reply:
Got it, sounds good 👍